AndroidHomeApplication/app/src/main/java/com/example/androidhomeapplication/ui/moviesList/FragmentMoviesList.kt

```diff
@@ -26,39 +29,96 @@ class FragmentMoviesList : Fragment(R.layout.fragment_movies_list) {
+                    viewModel.getCachedMovies()
+                    Log.d("FragmentMoviesList_LOG", "State.Empty")
+                }
+                is State.EmptyProgress -> {
+                    Log.d("FragmentMoviesList_LOG", "State.EmptyProgress")
+                }
+                is State.Data<*> -> {
+                    val movies = state.data as List<Movie>
+                    adapter.submitList(movies)
+
+                    if (viewModel.searchQuery.isEmpty()){
+                        viewModel.saveMoviesToCache(movies)
+                    }
+
+                    Log.d("FragmentMoviesList_LOG", "State.Data: page - ${state.page}")
+                }
+                is State.Refresh<*> -> {
+                    Log.d("FragmentMoviesList_LOG", "State.Refresh: page - ${state.page}")
+                }
+                is State.NewPageProgress<*> -> {
+                    Log.d("FragmentMoviesList_LOG", "State.NewPageProgress: page - ${state.page}")
+                }
+                is State.FullData<*> -> {
+                    Log.d("FragmentMoviesList_LOG", "State.FullData: page - ${state.page}")
+                }
+            }
+        }
+
+        viewModel.cachedMovies.observe(viewLifecycleOwner) { movies ->
+            if (viewModel.viewState.value == State.Empty){
+                adapter.submitList(movies)
+                viewModel.refresh()
+                Log.d("FragmentMoviesList_LOG", "Show Cached data and call refresh")
+            }
+        }
+
         binding.textInputLayout.textInputEditText.textChanges()
-            .debounce(500)
+            .debounce(300)
             .map { charSequence -> charSequence?.toString() }
             .distinctUntilChanged()
             .map { value ->
-                viewModel.loadData(value)
-                pagingAdapter.refresh()
+                if (viewModel.searchQuery != value) {
+                    adapter.submitList(emptyList())
+                    viewModel.restart(value ?: "")
+                }
             }
             .launchIn(viewLifecycleOwner.lifecycleScope)
 
-        viewLifecycleOwner.lifecycleScope.launch {
-            viewModel.movieItems.collectLatest { data ->
-                pagingAdapter.submitData(data)
+        binding.cinemaRv.addOnScrollListener(object : RecyclerView.OnScrollListener() {
+            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
+                if (dy > 0 && recyclerView.layoutManager != null) {
+                    val layoutManager = recyclerView.layoutManager!!
+
+                    val visibleItemCount = layoutManager.childCount
+                    val totalItemCount = layoutManager.itemCount
+                    val pastVisibleItem =
+                        (recyclerView.layoutManager as GridLayoutManager).findFirstVisibleItemPosition()
+                    if (!isLoading && (visibleItemCount + pastVisibleItem) >= totalItemCount) {
```

Comment:
!isLoading
should be done inside vm
just check:
some states logically contains this flag